[Merged by Bors] - feat(order/complete_well_founded): characterise well-foundedness for complete lattices

ocfnash opened the pull request:
---
- [x] depends on: #5446

Commit:
feat(src/order/complete_well_founded): characterise well-foundedness for complete lattices

Commit:
Generalise `finite_subset_is_bounded`

Commit:
Improve proofs

Still quite a bit of work to do but proofs reasonable now.

Commit:
Generalise `nat_rel_embedding_map_sup` and improve module doc

Commit:
Add lemma `Inf_eq_top_nonempty_eq_singleton`

Just for completeness since I have already added `Sup_eq_bot_nonempty_eq_singleton`

Commit:
Update src/data/finset/lattice.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Commit:
Update src/order/complete_well_founded.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

Commit:
Update src/order/complete_well_founded.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

Commit:
Drop 'rel_iso.monotone` in view of extant `strict_mono.monotone`.

Commit:
Make naming consistent after CR suggestions

Commit:
Naming suggestion from CR

Commit:
Add tfae lemma

Commit:
Add convenience `iff` lemmas for tfae result

Commit:
Add aliases for remaining implications

Commit:
Revert addition of new file `complete_well_founded.lean`

As per @urkud 's request, I am splitting this PR into two pieces:
  * One with all changes to existing files (this branch)
  * One with the proposed new file (a new branch).

Commit:
feat(order/complete_well_founded): characterise well-foundedness for complete lattices

Commit:
Merge branch 'master' into compact_complete_lattice_2

awainverse reviewed (commented):
This looks like you addressed all of my earlier comments, and the proofs are greatly simplified.
I have one suggestion, but consider it optional:

Since I last looked at this PR (or perhaps its predecessor?) I have been reading about lattice theory, and I'd like to see a definition of compact elements in a lattice (your statements here are essentially that every element is compact). If compact elements are defined, you can rephrase a lot of this file about compact elements (for instance, well-foundedness actually follows just from every element being a Sup of compact elements). If you want to take a crack at that, great, if not, I can try to do that in a separate PR after this one merges.

awainverse reviewed (approved):
I think this PR is good. I will look into generalization to compact elements later.

jcommelin reviewed (approved):
LGTM

robertylewis commented on src/order/complete_well_founded.lean:
```diff
@@ -0,0 +1,116 @@
+
+/-- A compactness property for a complete lattice is that any `sup`-closed non-empty subset
+contains its `Sup`. -/
+def is_sup_closed_compact :=
```
```suggestion
def is_sup_closed_compact : Prop :=
```
for clarity

robertylewis commented on src/order/complete_well_founded.lean:
```diff
@@ -0,0 +1,116 @@
+
+/-- A compactness property for a complete lattice is that any subset has a finite subset with the
+same `Sup`. -/
+def is_Sup_finite_compact := ∀ (s : set α), ∃ (t : finset α), ↑t ⊆ s ∧ Sup s = t.sup id
```
```suggestion
def is_Sup_finite_compact : Prop := 
∀ (s : set α), ∃ (t : finset α), ↑t ⊆ s ∧ Sup s = t.sup id
```

robertylewis committed:
Apply suggestions from code review

ocfnash committed:
Merge branch 'master' into compact_complete_lattice_2

ocfnash committed:
Fix proof of `is_sup_closed_compact.well_founded` after merging master

This is necessary because `rel_iso.map_rel_iff` was flipped in https://github.com/leanprover-community/mathlib/pull/5556
